File: src/packages/manifest.go
package packages

import (
	"archive/zip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"slices"

	"github.com/pelletier/go-toml/v2"
	"nyeki.dev/nye/projects"
	"nyeki.dev/nye/utils"
)

// A nye.toml manifest.
type Manifest struct {
	Package  ManifestPackage  `toml:"package"`
	Exposes  ManifestExposes  `toml:"exposes"`
	Consumes ManifestConsumes `toml:"consumes"`
}

type ManifestPackage struct {
	Name    string `toml:"name" validate:"required,kebab-case,min=1,max=32"`
	Version string `toml:"version" validate:"required,semver,max=32"`
	Target  string `toml:"target" validate:"required,supported-target"`
}

type ManifestExposes struct {
	Bin []ManifestExposesBin `toml:"bin" validate:"unique=Name"` // Exposed symlinks on installation to the package's bundled binaries
	Env []ManifestExposesEnv `toml:"env" validate:"unique=Name"` // Exposed environment variables. They're not set to the system's env vars.
}

type ManifestExposesBin struct {
	Name string `toml:"name" validate:"required,min=1,max=32,safe-path-segment"`
	Path string `toml:"path" validate:"required,min=1,safe-path"`
}

type ManifestExposesEnv struct {
	Name  string `toml:"name" validate:"required,min=1,max=32,env-var-name"`
	Value string `toml:"value" validate:"required,max=1024"`
}

type ManifestConsumes struct {
	Env []ManifestConsumesEnv `toml:"name" validate:"unique=Name"`
}

type ManifestConsumesEnv struct {
	Name      string `toml:"name" validate:"required,min=1,max=32,env-var-name"`
	Separator string `toml:"separator" validate:"required,max=32"`
}

// Reads a manifest from a file in the system.
//
// Unlike `GetManifestFromZip()`, this function does not validate the manifest on system data.
// The validation is done on syntax and individual values only. This is due this function being
// intended to be used on already-installed packages, whose manifests have already been validated
// by `GetManifestFromZip()`.
//
// Arguments:
// * `path` - The path to the manifest file.
//
// Returns:
// * `Manifest` - The manifest, if no error occurred.
// * `error` - An error if occurred, `nil` otherwise.
func GetManifestFromFile(path string) (Manifest, error) {
	contents, err := os.ReadFile(path)
	if err != nil {
		return Manifest{}, fmt.Errorf("could not read manifest file: %v", err)
	}

	manifest := Manifest{}
	err = toml.Unmarshal(contents, &manifest)
	if err != nil {
		return Manifest{}, fmt.Errorf("could not parse manifest file: %v", err)
	}

	err = utils.ValidateStruct(&manifest)
	if err != nil {
		return Manifest{}, fmt.Errorf("manifest was not valid: %v", err)
	}

	return manifest, nil
}

func GetManifestFromZip(path string) (Manifest, error) {
	zipper, err := zip.OpenReader(path)
	if err != nil {
		return Manifest{}, fmt.Errorf("could not read package file: %v", err)
	}
	defer zipper.Close()

	for _, file := range zipper.File {
		info := file.FileInfo()
		if info.Name() == "package.toml" {
			file, err := file.Open()
			if err != nil {
				return Manifest{}, fmt.Errorf("could not open manifestin package file: %v", err)
			}
			defer file.Close()

			contents, err := io.ReadAll(file)
			if err != nil {
				return Manifest{}, fmt.Errorf("could not read manifest in package file: %v", err)
			}

			manifest := Manifest{}
			err = toml.Unmarshal(contents, &manifest)
			if err != nil {
				return Manifest{}, fmt.Errorf("could not parse manifest in package file: %v", err)
			}

			err = validateManifest(manifest, zipper)
			if err != nil {
				return Manifest{}, fmt.Errorf("manifest in package file was invalid: %v", err)
			}

			return manifest, nil
		}
	}

	return Manifest{}, fmt.Errorf("the package file did not contain a manifest")
}

func validateManifest(manifest Manifest, zipper *zip.ReadCloser) error {
	if err := utils.ValidateStruct(manifest); err != nil {
		return err
	}

	if err := validateManifestExposedBins(manifest, zipper); err != nil {
		return err
	}

	return nil
}

func validateManifestExposedBins(manifest Manifest, zipper *zip.ReadCloser) error {
	exposedNames := []string{}

	for _, bin := range manifest.Exposes.Bin {
		if slices.Contains(exposedNames, bin.Name) {
			return fmt.Errorf("there are two or more binaries exposed under the same name, `%v`, which is not allowed", bin.Name)
		}
		exposedNames = append(exposedNames, bin.Name)

		if bin.Path == "" {
			bin.Path = bin.Name
		}

		found := false

		for _, file := range zipper.File {
			info := file.FileInfo()

			if info.IsDir() {
				continue
			}

			if file.Name == filepath.Join("bin", bin.Path) {
				found = true
				break
			}
		}

		if !found {
			return fmt.Errorf("exposed binary `%v` at `%v` was not in package", bin.Name, bin.Path)
		}
	}

	return nil
}

func FromProjectManifest(manifest projects.Manifest, target string) Manifest {
	exposedBins := []ManifestExposesBin{}
	exposedEnvs := []ManifestExposesEnv{}
	consumedEnvs := []ManifestConsumesEnv{}

	exposedForTarget := manifest.Exposes.ForTarget(target)

	for _, bin := range exposedForTarget.Bin {
		if bin.Path == "" {
			bin.Path = bin.Name
		}

		exposedBins = append(exposedBins, ManifestExposesBin{
			Name: bin.Name,
			Path: bin.Path,
		})
	}

	for _, env := range exposedForTarget.Env {
		exposedEnvs = append(exposedEnvs, ManifestExposesEnv{
			Name:  env.Name,
			Value: env.Value,
		})
	}

	for _, env := range manifest.Consumes.Env {
		consumedEnvs = append(consumedEnvs, ManifestConsumesEnv{
			Name:      env.Name,
			Separator: env.Separator,
		})
	}

	result := Manifest{
		Package: ManifestPackage{
			Name:    manifest.Package.Name,
			Version: manifest.Package.Version,
			Target:  target,
		},
		Exposes: ManifestExposes{
			Bin: exposedBins,
			Env: exposedEnvs,
		},
		Consumes: ManifestConsumes{
			Env: consumedEnvs,
		},
	}

	return result
}

File: src/packages/manifest.rs
use std::collections::HashSet;
use std::path::PathBuf;

use anyhow::Context;
use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::semver::Semver;
use crate::targets::Target;
use crate::validation::Validate;
use crate::{projects, validation};

#[derive(Serialize, Deserialize)]
pub struct Manifest {
    pub package: ManifestPackage,

    #[serde(default, skip_serializing_if = "ManifestExposes::is_empty")]
    pub exposes: ManifestExposes,
}

fn collect_artifacts(
    artifacts: &[projects::manifest::ManifestExposesArtifact],
    target: &Target,
) -> Vec<ManifestExposesArtifact> {
    let mut result = Vec::new();
    for artifact in artifacts {
        if artifact.targets.contains(target) || artifact.targets.is_empty() {
            for link in &artifact.links {
                result.push(ManifestExposesArtifact {
                    link: link.clone(),
                    path: artifact.path.clone(),
                });
            }
        }
    }

    result
}

impl Manifest {
    /// Returns a package manifest from a project manifest.
    ///
    /// This function does not validate the contents of the project manifest, it's on the caller to
    /// ensure the manifest specified is valid. An invalid manifest input will produce an invalid
    /// manifest output.
    ///
    /// Arguments:
    /// * `manifest` - The project's manifest.
    /// * `target` - The target this package manifest is for.
    pub fn from_project_manifest(manifest: projects::Manifest, target: Target) -> Self {
        let bin = collect_artifacts(&manifest.exposes.bin, &target);
        let lib = collect_artifacts(&manifest.exposes.lib, &target);
        let env = {
            let mut result = vec![];

            for var in &manifest.exposes.env {
                if !var.targets.contains(&target) && !var.targets.is_empty() {
                    continue;
                }

                result.push(ManifestExposesEnv {
                    name: var.name.clone(),
                    value: var.value.clone(),
                });
            }

            result
        };

        Manifest {
            package: ManifestPackage {
                name: manifest.package.name,
                version: manifest.package.version,
                target,
            },
            exposes: ManifestExposes { bin, lib, env },
        }
    }
}

impl Validate for Manifest {
    fn validate(&self) -> anyhow::Result<()> {
        self.package
            .validate()
            .context("The package field of the manifest was invalid.")?;
        self.exposes
            .validate()
            .context("The manifest's configured exposed artifacts were invalid.")?;

        Ok(())
    }
}

#[derive(Serialize, Deserialize)]
pub struct ManifestPackage {
    pub name: String,
    pub version: Semver,
    pub target: Target,
}

impl Validate for ManifestPackage {
    fn validate(&self) -> anyhow::Result<()> {
        if !(1..=32).contains(&self.name.len()) {
            anyhow::bail!(concat!(
                "The package's name must be at least 1 character long and less or equal to 32 ",
                "characters long."
            ));
        }

        validation::is_kebab_case(&self.name)
            .context("The package name was not valid kebab case")?;

        if !self.target.is_supported() {
            anyhow::bail!("The configured target in the manifest is not supported by nye.");
        }

        Ok(())
    }
}

#[derive(Serialize, Deserialize, Default)]
pub struct ManifestExposes {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub bin: Vec<ManifestExposesArtifact>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub lib: Vec<ManifestExposesArtifact>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub env: Vec<ManifestExposesEnv>,
}

fn validate_artifacts(artifacts: &[ManifestExposesArtifact]) -> anyhow::Result<()> {
    let mut links = HashSet::new();
    for artifact in artifacts {
        artifact.validate().context(format!(
            "The exposed artifact `{}` was incorrectly configured.",
            artifact.path.display()
        ))?;

        if links.contains(&artifact.link) {
            anyhow::bail!(
                "Two or more exposed artifacts conflict on the linked name `{}`.",
                artifact.link
            )
        }

        links.insert(artifact.link.clone());
    }

    Ok(())
}

impl Validate for ManifestExposes {
    fn validate(&self) -> anyhow::Result<()> {
        validate_artifacts(&self.bin)
            .context("The exposed binaries were incorrectly configured.")?;
        validate_artifacts(&self.lib)
            .context("The exposed libraries were incorrectly configured.")?;

        for var in &self.env {
            var.validate()
                .context("The exposed environment variables were incorrectly configured.")?;
        }

        Ok(())
    }
}

impl ManifestExposes {
    fn is_empty(&self) -> bool {
        self.bin.is_empty() && self.lib.is_empty()
    }
}

#[derive(Serialize, Deserialize)]
pub struct ManifestExposesArtifact {
    pub link: String,
    pub path: PathBuf,
}

impl Validate for ManifestExposesArtifact {
    fn validate(&self) -> anyhow::Result<()> {
        validation::is_safe_path(&self.path).context(format!(
            "The specified artifact path `{}` is not safe.",
            self.path.display()
        ))?;

        if !(1..=32).contains(&self.link.len()) {
            anyhow::bail!(
                "Linked names must be at least one character long and up to 32 characters long."
            );
        }

        validation::is_safe_path_component(&self.link).context(format!(
            "The specified artifact linked name `{}` was not a safe path component.",
            self.link
        ))?;

        Ok(())
    }
}

#[derive(Clone, Serialize, Deserialize)]
pub struct ManifestExposesEnv {
    pub name: String,
    pub value: String,
}

impl Validate for ManifestExposesEnv {
    fn validate(&self) -> anyhow::Result<()> {
        let regex = Regex::new("^[a-zA-Z0-9_]{1,32}$")
            .context("This is a bug. The hard-coded validation regex was invalid.")?;

        if !regex.is_match(&self.name) {
            anyhow::bail!("The package contained an environment variable with an invalid name.");
        }

        if self.value.len() > 512 {
            anyhow::bail!(
                "The package has an invalid variable exposed `{}`.",
                self.name
            );
        }

        Ok(())
    }
}
